safehttp/plugins/reportscollector/collector.go

```diff
@@ -0,0 +1,67 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package reportscollector
+
+import (
+	"encoding/json"
+
+	"github.com/google/go-safeweb/safehttp"
+)
+
+// Report represents a CSP violation report as specified by
+// https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy-Report-Only#Violation_report_syntax
+type Report struct {
+	BlockedURI         string `json:"blocked-uri"`
+	Disposition        string `json:"disposition"`
+	DocumentURI        string `json:"document-uri"`
+	EffectiveDirective string `json:"effective-directive"`
+	OriginalPolicy     string `json:"original-policy"`
+	Referrer           string `json:"referrer"`
+	ScriptSample       string `json:"script-sample"`
+	StatusCode         int    `json:"status-code"`
+	ViolatedDirective  string `json:"violated-directive"`
+}
+
+type report struct{}
+
+// Collector is a reports collector.
+type Collector interface {
+	Collect(Report)
+}
```

Comment:
Consider the ergonomics trade-off of defining an interface vs making a func type (`type Collector func(...)`).